Add an uninstall script

WeetHet opened the pull request:
Closes #14306

This looks at what #16660 did and install.sh script as a base for the uninstall.sh script. The script is bundled with the cli by default unless the cli/no-bundled-uninstall feature is selected which is done, so package managers could build zed without bundling a useless feature and increasing binary size.

I don't have capabilities to test this right now, so any help with that is appreciated.

Release Notes:

- Added an uninstall script for Zed installations done via zed.dev. To uninstall zed, run `zed --uninstall` via the CLI binary.

WeetHet committed:
Add an uninstall script

closes #14306

This looks at what #16660 did and install.sh script
as a base for the uninstall.sh script. The script is
bundled with the cli by default unless the cli/no-bundled-uninstall
feature is selected which is done, so package managers
could build zed without bundling a useless feature and
increasing binary size.

I don't have capabilities to test this right now,
so any help with that is appreciated.

maxdeviant commented:
> The script is bundled with the cli by default unless the cli/no-bundled-uninstall feature is selected which is done, so package managers could build zed without bundling a useless feature and increasing binary size.

This seems like it amounts to pretty insignificant savings for the additional complexity it incurs. I think we should always include it.

WeetHet commented:
> > The script is bundled with the cli by default unless the cli/no-bundled-uninstall feature is selected which is done, so package managers could build zed without bundling a useless feature and increasing binary size.
> 
> This seems like it amounts to pretty insignificant savings for the additional complexity it incurs. I think we should always include it.

I don't know, @SomeoneToIgnore wanted it to be there

SomeoneToIgnore commented on script/uninstall.sh:
```diff
@@ -0,0 +1,84 @@
+    fi
+
+    # Remove the binary symlink
+    rm -f "$HOME/.local/bin/zed"
```
I think we would also need to deal with the `~/Library/Application\ Support/Zed` directory, but it's a bit trickier, as multiple Zed versions may store things there:

```
Library/Application Support/Zed ❯ l .
total 16
drwxr-xr-x@  12 someonetoignore  staff   384B Oct  3 17:38 ./
drwx------+ 169 someonetoignore  staff   5.3K Nov 23 02:16 ../
-rw-r--r--@   1 someonetoignore  staff   6.0K Mar 26  2024 .DS_Store
drwxr-xr-x    3 someonetoignore  staff    96B Jul 13 03:12 copilot/
drwxr-xr-x    7 someonetoignore  staff   224B Sep 19 22:23 db/
drwxr-xr-x    3 someonetoignore  staff    96B Jun 14 01:37 docs/
drwxr-xr-x    6 someonetoignore  staff   192B Nov  6 01:15 extensions/
drwxr-xr-x   14 someonetoignore  staff   448B Nov 16 18:02 languages/
drwxr-xr-x@   6 someonetoignore  staff   192B Sep 24 23:20 node/
drwxr-xr-x    6 someonetoignore  staff   192B May 30 17:54 prettier/
drwxr-xr-x    3 someonetoignore  staff    96B Oct  3 17:38 remote_servers/
drwxr-xr-x    3 someonetoignore  staff    96B Jul 24 02:25 supermaven/
```

Technically, removing most of these is fine, as other Zed versions will redownload the needed things.
It's not great, esp. if the person is behind the VPN or offline, but seems like a better thing that not removing anything at all (as usually, there's only one version Zed users have installed).

The only notable thing is the `db` dir:
```
Library/Application Support/Zed ❯ l db
total 0
drwxr-xr-x   7 someonetoignore  staff   224B Sep 19 22:23 ./
drwxr-xr-x@ 12 someonetoignore  staff   384B Oct  3 17:38 ../
drwxr-xr-x   5 someonetoignore  staff   160B May 30 18:27 0-dev/
drwxr-xr-x   5 someonetoignore  staff   160B Sep 19 22:23 0-global/
drwxr-xr-x   5 someonetoignore  staff   160B May 30 17:54 0-nightly/
drwxr-xr-x   5 someonetoignore  staff   160B May 30 21:23 0-preview/
drwxr-xr-x   5 someonetoignore  staff   160B May 30 21:23 0-stable/
```

we should only remove the one corresponding to the current version, and only remove `db` if it's empty afterwards.

-------------

Similar thing has to be done for Linux, it's in `~/.local/share/zed/` and besides all that have `zed-$VERSION.sock` files that have to follow the same logic.

WeetHet replied:
What I wanna do is the following:
```bash
db_dir="$HOME/Library/Application Support/Zed/db"
if [ -d "$db_dir/0-$db_suffix" ]; then
    rm -rf "$db_dir/0-$db_suffix"
    # If only global db remains, remove entire Zed directory
    if [ "$(ls -1 "$db_dir" | grep -v global | wc -l)" -eq 0 ]; then
        rm -rf "$HOME/Library/Application Support/Zed"
    fi
fi
```
So, by using database as an indicator, we can pretty accurately determine whether there are any more installations left

WeetHet replied:
Also, what is the .sock file name? Is it just `zed-{nightly,preview,stable}.sock`? Is it in the ~/.local/share/zed?

SomeoneToIgnore replied:
Yes to both.

SomeoneToIgnore commented on script/uninstall.sh:
```diff
@@ -0,0 +1,84 @@
+
+    # Remove the binary symlink
+    rm -f "$HOME/.local/bin/zed"
+}
```
Similarly, there's `~/.config/zed/` on both Linux and macOS, which I'm not certain what to do with, if someone has more than one version installed.

I wonder if we should prompt for deleting that and let the user to decide?

WeetHet replied:
Yeah, sure, would the following work:
```bash
printf "Do you want to keep your Zed preferences? [Y/n] "
read -r response
case "$response" in
    [nN]|[nN][oO])
        rm -rf "$HOME/.config/zed"
        echo "Preferences removed."
        ;;
    *)
        echo "Preferences kept."
        ;;
esac
```

SomeoneToIgnore replied:
Come to think of it, maybe we can check if there are other versions, via https://github.com/zed-industries/zed/pull/21213#discussion_r1860559294 , and only remove it if it's the only version in the system, being removed.

SomeoneToIgnore commented on script/uninstall.sh:
```diff
@@ -0,0 +1,84 @@
+    # Remove the binary symlink
+    rm -f "$HOME/.local/bin/zed"
+}
+
```
Last but not least, after ssh remoting there's `~/.zed_server/` which we have to remove.
That is "fine" to remove without asking, as it will be redownloaded for other versions.

WeetHet replied:
I thought it is on the remote, not the host? I've used remoting and I don't have that directory

SomeoneToIgnore replied:
You never know, one can ssh into itself, for instance.

notpeter replied:
I'm with @SomeoneToIgnore. We should cleanup `~/.zed_server` too.

SomeoneToIgnore commented on script/uninstall.sh:
```diff
@@ -0,0 +1,84 @@
+    elif [ "$platform" = "Linux" ]; then
+        platform="linux"
+    else
+        echo "Unsupported platform $platform"
```
Let's support FreeBSD here also, as it had been recently semi-supported in https://github.com/zed-industries/zed/pull/20480

WeetHet replied:
The installation script doesn't support freebsd and I'm not ready to take any responsibility for it working correctly

SomeoneToIgnore reviewed (commented):
Thank you, amazing to see something long awaited coming in, and without without `curl | sh`.

Overall, looks good, but I cannot miss the opportunity to emphasize how bad https://github.com/zed-industries/zed/pull/16660 would work given that it's outdated enough to miss at least 3 dirs by this point.

I am not totally sure that zed cli embedding is worth it, as it turned out we'd better prompt about particular deletions, and may benefit from extra logging in the script overall, so people see which directories are cleaned.

Another (small) concern is the fact that people still tend to confuse zed-cli with zed binary, so that option might get unnoticed.

If you think that extra prompting and logging is doable via zed-cli, let's move on with this approach.
Otherwise, let's put this `uninstall.sh` inside `~/.local/zed-$VERSION.app/` (and whatever macOS analogue) and package it with `bundle-{linux,mac}.sh` scripts.

WeetHet committed:
Remove more files

This removes zed database for the corresponding
channel as well as a bunch of directories on macos
and a socket file on linux. This also adds support
for removing preferences, asking the user about it
if no installations are left

WeetHet committed:
Put cfg guard directly onto the if statement

Moves the cfg guard out from the block containing the
uninstall if statement onto the if statement itself,
reducing nesting level by one

WeetHet committed:
Also remove ~/.zed_server

WeetHet committed:
Merge branch 'main' into uninstall-script

WeetHet committed:
Detect zed channel

WeetHet committed:
Don't make UNINSTALL_SCRIPT lazy

WeetHet committed:
Enable no-bundled-uninstall if ZED_UPDATE_EXPLANATION is set

SomeoneToIgnore reviewed (approved):
Tested on both systems with various installations and now it seems to do what's needed.

I am not super fond of all the feature fiddling, but like the idea of reusing zed-cli for this purpose, so let's see how it works with the other users.

Thank you a lot for writing all that.

antaz commented:
`--uninstall` is a bit unusual for CLI tools, but i like when Zed breaks the rules.
